parser/README.md

```diff
@@ -0,0 +1,134 @@
+| SQL Engine | Parser  |
+|------------|---------|
+| Beam       | Calcite |
+| Flink      | Calcite |
+| Storm      | Calcite |
+| Hive       | Hplsql.g4 |
+| MySQL      | sql_parse.{h,cc} |
+
+It is notable that MySQL doesn't provide sufficient documentation on how to call its parser, so we call a Go implementation of MySQL's SQL parser by PingCap https://github.com/pingcap/parser instead. For proprietary engines, we have a plan to ask the licensing of their parsers.
+
+
+## The External Parser Abstraction
+
+The above example program calls `sql_engine.Parse` twice before calling `SQLFlow.Parse`.  In practice, because Calcite parser is in Java, SQLFlow is a Go program, to enable SQLFlow calling Calcite parser, we have to wrap Calcite parser up into a gRPC server.  It is time-consuming to make an RPC call, so we pack the two calls to `sql_engine.Parse` into one:
+
+```protobuf
+message CalciteParserRequest {
+  string query = 1;
+}
+
+message CalciteParserReply {
+  string sql = 1;
+  string extension = 2;
+  string error = 3;
+}
+
+service CalciteParser {
+  rpc Parse (CalciteParserRequest) returns (CalciteParserReply) {}
+}
+```
+
+Some external parsers are in Go and don't need RPC.  For example, TiDB parser is in Go, and HiveQL parser is a set of AntLR grammar rules, and AntLR can generate Go code.  However, we would unify the interface to external parsers:
+
+```go
+func external_parser(kind, sql string) (idx int, err error) {
+   switch(kind) {
+   case 'calcite':
+       r := grpc.Call("CalciteParser.Parse", CalciteParserRequest{sql})
+       return r.GetIndex(), r.GetError()
+   case 'mysql':
+       return tidb.Parse(sql)
+   ...
+   }
+}
+```
+
+Now we can change the above exmaple program to call `external_parser` and `SQLFlow.Parse`.
+
+```go
+func Parse(sql string) error {
+    i, e := external_parser(sql)
+    if e != nil {
+      return e
+    }
+    
+    if i != -1 {
+      return SQLFlow.Parse(sql[i,:]) // Parse the right part.
+    }
+    return nil
+}
+```
+
+The above design of `external_parser` returns `idx int`, which separate the input string `sql` into the left part, the "standard" SQL, and the right part, the SQLFlow extension.  In reality, we need it to return more than `idx`.
+
+
+## Get Prepared for Logical Verification
+
+The parser verifies the syntax of the input.  SQLFlow also needs to check the logic.  For example, have a look at the following SQLFlow statement
+
+```sql
+SELECT a, b FROM t1 TRAIN DNNClassifier COLUMN c LABEL b;
+```
+
+There is no syntax error in the above example; however, it has a logical mistake -- the column `c` is undefined.
+
+In the design of SQLFlow, we do verification after parsing.  The verifier would like to know the field and table named mentioned in the statement, which is supposed to be returned by `external_parser` in addition to `idx`. Both TiDB parser and Calcite parser return an abstract syntax tree; it seems that we can traverse the tree and find out the field and table names.
```

Comment:
Just curious, in order to retrieve the `abstract syntax tree`, the RPC reply like `CalciteParserReply` should return a tree structure that can be serialized by protobuf, maybe can be a [protobuf map](https://developers.google.com/protocol-buffers/docs/proto3#maps)

Reply:
We might make the RPC server traverse the AST and return two lists -- fields and tables.

Reply:
I see, Thanks!